### PR TITLE
Infra/color picker to ts

### PR DESCRIPTION
**Pull request by Inbal-Tish**
## Description
Migrating to TypeScript:
1) ColorPicker
2) ColorSwatch
3) ColorPalette
4) ColorPickerDialog

## Changelog
Migrating to TypeScript:
1) ColorPicker
2) ColorSwatch
3) ColorPalette
4) ColorPickerDialog

### CODE REVIEW

**Comment by M-i-k-e-l** [comment]
Any chance this can be split into smaller pieces? :shaking:

**Comment by Inbal-Tish** [comment]
> Any chance this can be split into smaller pieces? :shaking:

There are 9 files for generated types that you don't need to review and another 2 are also related to types. If you still need smaller PRs I can do it when I'll get back from vacation...

**Comment by M-i-k-e-l** [comment]
I'd really appreciate it if you could split it 🙏

**Comment by Inbal-Tish** [comment]
@M-i-k-e-l  All these components are being used by the ColorPicker component. How do you suggest I'll split the PR?

**Comment by M-i-k-e-l** [comment]
> @M-i-k-e-l All these components are being used by the ColorPicker component. How do you suggest I'll split the PR?

You can start with a used component, since you can use a ts component in js component.
Once the composed components are done the composing component can be refactored.

File: src/components/colorPicker/ColorPickerDialog.tsx [hunk 1]
```diff
@@ -197,7 +225,7 @@ class ColorPickerDialog extends PureComponent {
   };
 
   renderHeader() {
-    const {useCustomTheme, accessibilityLabels} = this.props;
+    const {buttonColor, accessibilityLabels} = this.props;
```
Comment (ethanshar) [inline src/components/colorPicker/ColorPickerDialog.tsx:228]: `buttonColor` is too vague.. what about `confirmButtonColor`?

File: src/components/colorPicker/ColorPickerDialog.tsx [hunk 2]
```diff
@@ -22,45 +70,17 @@ const KEYBOARD_HEIGHT = 216;
  * @extends: Dialog
  * @example: https://github.com/wix/react-native-ui-lib/blob/master/demo/src/screens/componentScreens/ColorPickerScreen.js
  */
-class ColorPickerDialog extends PureComponent {
+class ColorPickerDialog extends PureComponent<Props, State> {
   static displayName = 'ColorPickerDialog';
```
Comment (ethanshar) [inline src/components/colorPicker/ColorPickerDialog.tsx:74]: ColorPickerDialog is an internal ColorPicker component. 
If you want its props (interface) to be included in ColorPicker docs you'll need to set its displayName to be `ColorPicker` like the main component.

File: src/components/colorPicker/ColorPickerDialog.tsx [hunk 3]
```diff
@@ -154,15 +182,16 @@ class ColorPickerDialog extends PureComponent {
...
     const hex = this.getHexString(color);
     const text = this.getColorValue(hex);
     this.setState({color, text, valid: true});
   }
 
   resetValues() {
-    const color = Colors.getHSL(this.props.initialColor);
-    const text = this.getColorValue(this.props.initialColor);
+    const {initialColor} = this.props;
+    const color = initialColor && Colors.getHSL(initialColor);
```
Comment (ethanshar) [inline src/components/colorPicker/ColorPickerDialog.tsx:193]: move undefined check to the getHSL method

File: src/components/colorPicker/ColorPickerDialog.tsx [hunk 4]
```diff
@@ -264,27 +292,30 @@ class ColorPickerDialog extends PureComponent {
...
               autoCorrect={false}
-              autoComplete={'off'}
+              autoCompleteType={'off'}
               autoCapitalize={'characters'}
               // keyboardType={'numbers-and-punctuation'} // doesn't work with `autoCapitalize`
               returnKeyType={'done'}
               enablesReturnKeyAutomatically
               onFocus={this.onFocus}
-              accessibilityLabel={accessibilityLabels.input}
+              accessibilityLabel={_.get(accessibilityLabels, 'input')}
```
Comment (ethanshar) [inline src/components/colorPicker/ColorPickerDialog.tsx:318]: Why not? 
```suggestion
              accessibilityLabel={accessibilityLabels?.input}
```

File: src/components/colorPicker/ColorPickerDialog.tsx [hunk 5]
```diff
@@ -300,7 +331,7 @@ class ColorPickerDialog extends PureComponent {
       <Dialog
         visible={visible} //TODO: pass all Dialog props instead
         width="100%"
-        height={null}
+        height={undefined}
         bottom
```
Comment (ethanshar) [inline src/components/colorPicker/ColorPickerDialog.tsx:335]: I remember passing `null` to Dialog's height has different outcome than `undefined`. Why did you change it?

File: src/index.ts [hunk 6]
```diff
@@ -30,6 +30,10 @@ export {default as TouchableOpacity, TouchableOpacityProps} from './components/t
 export {default as Button, ButtonPropTypes, ButtonProps, ButtonSize, ButtonAnimationDirection} from './components/button';
 export {default as Checkbox, CheckboxPropTypes, CheckboxProps} from './components/checkbox';
 export {default as Chip, ChipPropTypes, ChipProps} from './components/chip';
+export {default as ColorPicker, ColorPickerProps} from './components/colorPicker';
+export {default as ColorPickerDialog, ColorPickerDialogProps} from './components/colorPicker/ColorPickerDialog';
```
[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply src/index.ts:35]: Yes. Will remove

File: src/components/colorPicker/ColorPalette.tsx [hunk 7]
```diff
@@ -1,17 +1,70 @@
...
+   */
+  animatedIndex?: number;
+  /**
+   * Invoked once when value changes by selecting one of the swatches in the palette
+   */
+  onValueChange?: (value: string, options: object) => void;
+  style?: StyleProp<ViewStyle>;
+  testID?: string;
+}
+export type ColorPaletteProps = Props;
```
[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorPalette.tsx:59]: Because I like to use 'Props' as the type in the internal use and separate it from the exported usage of 'ColorPaletteProps'.

File: src/components/colorPicker/ColorPickerDialog.tsx [hunk 8]
```diff
@@ -1,19 +1,67 @@
...
+   * Ok (v) button color
+   */
+  buttonColor?: string,
+  accessibilityLabels?: {
+    dismissButton?: string,
+    doneButton?: string,
+    input?: string
+  };
+}
+export type ColorPickerDialogProps = Props;
```
[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorPickerDialog.tsx:57]: Same here

File: src/components/colorPicker/ColorSwatch.tsx [hunk 9]
```diff
@@ -1,16 +1,43 @@
...
+  animated?: boolean;
+  /**
+   * onPress callback
+   */
+  onPress?: (value: string, options: object) => void;
+  index?: number;
+  style?: StyleProp<ViewStyle>;
+  testID?: string;
+}
+export type ColorSwatchProps = Props;
```
[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorSwatch.tsx:39]: Same here

File: src/components/colorPicker/ColorSwatch.tsx [hunk 10]
```diff
@@ -143,10 +150,9 @@ export default class ColorSwatch extends PureBaseComponent {
         onPress={this.onPress}
         style={[this.styles.container, style]}
         onLayout={this.onLayout}
-        accessibilityLabel={Colors.getColorName(color)}
-        accessibilityStates={this.props.selected ? ['selected'] : []}
+        {...this.getAccessibilityInfo()}
       >
-        {Colors.isTransparent(color) && (
+        {color && Colors.isTransparent(color) && (
```
[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorSwatch.tsx:155]: Correct. Will remove the check

File: src/components/colorPicker/ColorSwatch.tsx [hunk 11]
```diff
@@ -108,30 +106,39 @@ export default class ColorSwatch extends PureBaseComponent {
...
     this.layout = event.nativeEvent.layout;
   };
 
   renderContent() {
-    const {style, color, onPress, onValueChange, ...others} = this.getThemeProps();
+    const {style, color, onPress, ...others} = this.props;
     const {isSelected} = this.state;
-    const Container = onPress || onValueChange ? TouchableOpacity : View;
-    const tintColor = this.getTintColor(color);
+    const Container = onPress ? TouchableOpacity : View;
```
[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorSwatch.tsx:140]: It's not changing functionality, there is no `onValueChange` prop on ColorSwatch component...

File: src/components/colorPicker/ColorSwatch.tsx [hunk 12]
```diff
@@ -108,30 +106,39 @@ export default class ColorSwatch extends PureBaseComponent {
...
+    const tintColor = color && this.getTintColor(color);
```
[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorSwatch.tsx:141]: Done. Thanks

File: src/components/colorPicker/ColorSwatch.tsx [hunk 13]
```diff
@@ -94,7 +92,7 @@ export default class ColorSwatch extends PureBaseComponent {
     ]).start();
   }
 
-  animateCheckmark(newValue) {
+  animateCheckmark(newValue: any) {
```
[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorSwatch.tsx:95]: Yes. Done

File: src/components/colorPicker/ColorSwatch.tsx [hunk 14]
```diff
@@ -22,59 +49,30 @@ export const SWATCH_SIZE = DEFAULT_SIZE;
...
       this.animateCheckmark(this.props.selected);
     }
   }
 
-  generateStyles() {
-    this.styles = createStyles(this.getThemeProps());
-  }
-
-  animateSwatch(newValue) {
+  animateSwatch(newValue: any) {
```
[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorSwatch.tsx:75]: Yes. Done

File: src/components/colorPicker/ColorPickerDialog.tsx [hunk 15]
```diff
@@ -22,45 +70,17 @@ const KEYBOARD_HEIGHT = 216;
...
   static defaultProps = {
     initialColor: Colors.dark80
   };
 
-  constructor(props) {
+  constructor(props: Props) {
     super(props);
 
-    const color = Colors.getHSL(props.initialColor);
+    const color = props.initialColor && Colors.getHSL(props.initialColor);
```
[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorPickerDialog.tsx:83]: I didn't want to change the implementation of the method, as other components use it. Maybe it's proper to a separate PR?

File: src/components/colorPicker/ColorPickerDialog.tsx [hunk 16]
```diff
@@ -99,52 +125,54 @@ class ColorPickerDialog extends PureComponent {
   };
 
   setFocus = () => {
-    if (this.textInput) {
-      this.textInput.focus();
+    if (this.textInput && this.textInput.current) {
+      this.textInput.current.focus();
     }
```
[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorPickerDialog.tsx:130]: Yes. Thanks.

File: src/components/colorPicker/ColorPickerDialog.tsx (hunk 1 above)
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorPickerDialog.tsx:228]: Ok. Changed to 'doneButtonColor'.

File: src/components/colorPicker/ColorPickerDialog.tsx (hunk 2 above)
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorPickerDialog.tsx:74]: I didn't know that. Why is a simple `export` and `extends` not enough?

Comment (Inbal-Tish) [reply src/components/colorPicker/ColorPickerDialog.tsx:74]: Done

File: src/components/colorPicker/ColorPickerDialog.tsx (hunk 3 above)
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorPickerDialog.tsx:193]: As I mentioned before, I don't think changing this implementation should be in this PR scop as this is not a ColorPickerDialog's method but Color's, which other components use.

File: src/components/colorPicker/ColorPickerDialog.tsx (hunk 5 above)
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorPickerDialog.tsx:333]: It's actually meaningless now and I'll remove the prop entirely. If I remember correctly, the point of passing null was to get the dialog to take the content height. Now we can just not pass height. The only way to change the Dialog's height is from 'dialogProps', which should be deprecated.

File: src/components/colorPicker/ColorPickerDialog.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/components/colorPicker/ColorPickerDialog.tsx:83]: Changing `Colors.getHSL` color argument to optional doesn't affect anything and will save you from checking for undefined value.

File: src/components/colorPicker/ColorPickerDialog.tsx (hunk 3 above)
Comment (ethanshar) [reply src/components/colorPicker/ColorPickerDialog.tsx:191]: See my previous comment, it's a small change in `Colors`. No reason not to include it.

File: src/components/colorPicker/ColorPickerDialog.tsx (hunk 5 above)
Comment (ethanshar) [reply src/components/colorPicker/ColorPickerDialog.tsx]: Ok great.

File: src/components/colorPicker/ColorPalette.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/components/colorPicker/ColorPalette.tsx:59]: We usually separate the names when there's a difference between what we export and what the internal component is expecting to get. 
In this case they are the same. 
Personally I feel it's more readable to have `ColorPaletteProps` than `Props`. 

Anyway, it's fine either way. You decide

[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorPalette.tsx:59]: I know. I just think it's cleaner to separate and use `Props` in the code instead of the full name.

File: src/components/colorPicker/ColorPickerDialog.tsx (hunk 3 above)
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorPickerDialog.tsx:191]: Ok

Comment (Inbal-Tish) [reply src/components/colorPicker/ColorPickerDialog.tsx]: Done

File: src/components/colorPicker/ColorPickerDialog.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply src/components/colorPicker/ColorPickerDialog.tsx]: Done

**Review by ethanshar — APPROVED** [review]
